doc/smart-switch/smart-switch-database-architecture/smart-switch-database-design.md

```diff
@@ -0,0 +1,262 @@
+| ---- | --------------------------------- |
+| NPU  | Network Processing Unit           |
+| DPU  | Data Processing Unit              |
+| DB   | Database                          |
+| GNMI | gRPC Network Management Interface |
+
+### Overview
+
+The Smart Switch comprises two integral components: the Network Processing Unit (NPU) and the Data Processing Unit (DPU), both operating on the SONiC OS. The database stack encompasses the entire database infrastructure for both the NPU and DPU. However, due to memory limitations on the DPU, certain overlayer objects, such as DASH objects, are stored in the NPU.
+
+In addition, dedicated database containers are maintained in the NPU for each DPU, serving the purpose of resource management within the Smart Switch architecture. This separation allows for efficient handling of database-related operations and ensures optimal utilization of resources across the entire Smart Switch.
+
+
+### Requirements
+
+- All databases, including those on both NPU and DPU, must be accessible through the GNMI server.
+- Each DPU database instance on the NPU is associated with a unique TCP port and domain Unix socket path.
+- All DPU database instances on the NPU will be bound to the IP address of the midplane bridge.
+- All database instances on the NPU share the same network namespace to facilitate seamless communication.
+- DPUs can access their respective overlay database instances using the IP of the midplane bridge and a pre-assigned unique TCP port.
+
+### Architecture Design
+
+![smart-switch-database-architecture](smart-switch-database-architecture.png)
+
+#### Database services
+
+In this section, the focus is on illustrating the maintenance of DPU overlay databases within the NPU. It's essential to note that the traditional database services of both NPU and DPU remain unchanged and do not necessitate further design modifications.
+
+The management of DPU overlay databases within the NPU is orchestrated through existing SONiC database services. The daemon, named "featured," retains the responsibility for initiating, terminating, enabling, and disabling the DPU overlay database services. This interaction is facilitated using the systemctl tool.
+
+To determine the DPU number, the "featured" daemon should leverage the platform API. However, for the sake of implementation simplicity, the DPU number is extracted directly from the platform_env.conf file firstly.
+
+``` shell
+cat /usr/share/sonic/device/$PLATFORM/platform_env.conf
+NUM_DPU=2
+```
+
+To align with the established multi-ASIC design in SONiC, a new field, `"has_per_dpu_scope": "True"``, is introduced in the database feature table within config_db.json. This field plays a crucial role in ensuring that each DPU database instance is initiated within a dedicated database container. This design approach maintains consistency with SONiC's existing architecture while accommodating the specific requirements of DPU overlay databases.
```

Comment:
>ensuring [](http://example.com/codeflow?start=205&length=8)

Could you explain the meaning of `has_per_dpu_scope`? and how does it ensure "each DPU database instance is initiated within a dedicated database container" ?

Reply:
This is the same concept as the `has_per_asic_scope`. This field will be read by the `featured` to start database service container instances for each DPU.